src/foam/nanos/cron/CronSchedule.js

```diff
@@ -79,53 +89,69 @@ foam.CLASS({
         }
       ],
       type: 'Date',
-      javaCode:
-`Calendar next = Calendar.getInstance();
-next.setTime(from);
-next.add(Calendar.SECOND, 1);
-next.set(Calendar.MILLISECOND, 0);
+      javaCode: `
+var zone = ZoneId.systemDefault();
+if ( ! foam.util.SafetyUtil.isEmpty(getTimezone()) ) {
+  zone = ZoneId.of(getTimezone());
+}
 
-boolean dateFound = false;
-while ( next.get(Calendar.YEAR) < 3000 ) {
-  if ( getMonth() >= 0 && next.get(Calendar.MONTH) != getMonth() - 1 ) {
-    next.add(Calendar.MONTH, 1);
-    next.set(Calendar.DAY_OF_MONTH, 1);
-    next.set(Calendar.HOUR_OF_DAY, 0);
-    next.set(Calendar.MINUTE, 0);
-    next.set(Calendar.SECOND, 0);
-    continue;
-  }
-  if ( ( getDayOfMonth() >= 0 && next.get(Calendar.DAY_OF_MONTH) != getDayOfMonth() ) ||
-      ( getDayOfWeek() >= 0 && next.get(Calendar.DAY_OF_WEEK) != getDayOfWeek() + 1) ) {
-    next.add(Calendar.DAY_OF_MONTH, 1);
-    next.set(Calendar.HOUR_OF_DAY, 0);
-    next.set(Calendar.MINUTE, 0);
-    next.set(Calendar.SECOND, 0);
-    continue;
+LocalDateTime last = null;
+if ( from == null ) {
+  last = LocalDate.now(zone).atStartOfDay();
+} else {
+  last = LocalDateTime.ofInstant(from.toInstant(), zone);
+}
+var time = last;
+
+// test if component needs to be bumped to the next hour, minute,...
+if ( getHour() > -1 ) {
+  if ( time.getHour() >= getHour() &&
+       ! time.isAfter(last) ) {
```

Comment:
time = last so not sure you need this check

Reply:
Yes, this particular test is not needed. 
Presently all the test conditions are the same and I feel I'd want to comment why the first one is different. 
I would like to leave it, but you make the call.